app/dao/notifications_dao.py

```diff
@@ -267,47 +356,63 @@ def update_notification_status_by_id(
     if not notification.sent_by and sent_by:
         notification.sent_by = sent_by
 
-    if is_feature_enabled(FeatureFlag.NOTIFICATION_FAILURE_REASON_ENABLED) and status_reason:
-        notification.status_reason = status_reason
-
     return dao_update_notification_by_id(
         notification_id=notification.id,
         status=status,
-        status_reason=notification.status_reason,
-        sent_by=notification.sent_by,
+        status_reason=status_reason,
+        sent_by=sent_by,
     )
 
 
 @statsd(namespace='dao')
-def update_notification_delivery_status(
+def dao_update_notification_delivery_status(
     notification_id: UUID,
+    notification_type: str,
     new_status: str,
-    new_status_reason: str = None,
-) -> None:
-    """
-    Update a notification's delivery status.
+    new_status_reason: str | None,
+    segments_count: int,
+    cost_in_millicents: float,
+) -> Notification:
+    """Update an SMS notification delivery status.
 
     Based on the desired update status, a query is constructed that cannot put the notification in an incorrect state.
 
-    Arguments:
-        notification_id (UUID): Notification id,
-        new_status (str): Status to update the notification to,
+    Args:
+        provider (str): The provider being used
+        notification_id (UUID): The id to update
+        new_status (str): The new status being set
+        new_status_reason (str, optional): The new status reason beign set. Defaults to None.
+
+    Returns:
+        Notification: A Notification object
     """
 
-    current_app.logger.info('Update notification: %s to status: %s', notification_id, new_status)
-    stmt = _get_notification_status_update_statement(
-        notification_id=notification_id,
-        incoming_status=new_status,
-        incoming_status_reason=new_status_reason,
-    )
+    if notification_type == SMS_TYPE:
+        stmt = (
```

Comment:
I spent some time looking at this query to see if the efficient could be improved. After some review, I believe the current query is sufficient.

Reply:
Thank you for checking!